File: backend/src/services/stateService.js
import { config } from "../config.js";
import { readJson, writeJson } from "../utils/fileStore.js";

class StateService {
  load() {
    const state = readJson(config.statePath, {
      initialBudget: config.initialBudget,
      budgetRemaining: config.initialBudget,
      adminPaused: false,
      freezeReason: null,
      requestLog: [],
      activeQueue: [],
      claimOverrides: {}
    });

    if (typeof state.initialBudget !== "number") {
      state.initialBudget = config.initialBudget;
    }

    if (typeof state.budgetRemaining !== "number") {
      state.budgetRemaining = config.initialBudget;
    }

    if (!Array.isArray(state.requestLog)) {
      state.requestLog = [];
    }

    if (!Array.isArray(state.activeQueue)) {
      state.activeQueue = [];
    }

    if (!state.claimOverrides || typeof state.claimOverrides !== "object") {
      state.claimOverrides = {};
    }

    return state;
  }

  save(state) {
    writeJson(config.statePath, state);
  }

  get() {
    return this.load();
  }

  update(mutator) {
    const state = this.load();
    const nextState = mutator(structuredClone(state)) || state;
    this.save(nextState);
    return nextState;
  }
}

export const stateService = new StateService();

File: backend/flask_app/services/state_service.py
from copy import deepcopy
from flask import current_app

from ..utils.file_store import read_json, write_json


class StateService:
    def load(self):
        config = current_app.config
        state = read_json(
            config["STATE_PATH"],
            {
                "initialBudget": config["INITIAL_BUDGET"],
                "budgetRemaining": config["INITIAL_BUDGET"],
                "adminPaused": False,
                "freezeReason": None,
                "requestLog": [],
                "activeQueue": [],
                "claimOverrides": {},
            },
        )

        if not isinstance(state.get("initialBudget"), (int, float)):
            state["initialBudget"] = config["INITIAL_BUDGET"]

        if not isinstance(state.get("budgetRemaining"), (int, float)):
            state["budgetRemaining"] = config["INITIAL_BUDGET"]

        if not isinstance(state.get("requestLog"), list):
            state["requestLog"] = []

        if not isinstance(state.get("activeQueue"), list):
            state["activeQueue"] = []

        if not isinstance(state.get("claimOverrides"), dict):
            state["claimOverrides"] = {}

        return state

    def save(self, state):
        write_json(current_app.config["STATE_PATH"], state)

    def get(self):
        return self.load()

    def update(self, mutator):
        state = self.load()
        next_state = mutator(deepcopy(state))
        if next_state is None:
            next_state = state
        self.save(next_state)
        return next_state


state_service = StateService()
